src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java

```diff
@@ -0,0 +1,428 @@
+ * a pair of classes that follow this pattern. For more information and a higher level description
+ * of the component classes involved in a Spark-aware collector, see
+ * {@link org.broadinstitute.hellbender.tools.spark.pipelines.metrics.MetricsCollectorSpark}.
+ *
+ * To use this class:
+ *
+ * Subclass UNIT_COLLECTOR and implement a combine method that combines two UNIT_COLLECTORs. Subclass
+ * MultiLevelCollectorReducible, providing the UNIT_COLLECTOR-derived class as a type parameter, and implement
+ * a combineUnit method that delegates to the UNIT_COLLECTOR's combine method.
+ *
+ * Instantiate a MultiLevelCollectorReducible and call setup, which will will create the underlying classes
+ * that will handle the accumulation level logic. Pass all reads you wish to collect data against to
+ * MultiLevelCollectorReducible via the acceptRecord method. Call finish when all reads have been processed.
+ * When multiple MultiLevelCollectorReducible have been created in parallel, the set of collectors can be
+ * reduced using the combine method (which will aggregate like per-unit collectors and call the combineUnit
+ * method).
+ *
+ * Extend MultiLevelCollectorReducible and implement makeArg and makeChildCollector
+ * You will need to make a class that extends PerUnitMetricCollector.  This class should do the work of keeping
+ * track of values for one specific "accumulation unit" (e.g. for one library, or for one read group depending on
+ * what levels you are accumulating at).
+ *
+ * If a record has any expensive calculations to be done (that don't need to be done differently depending
+ * on what sample/library/read group the read is for) then create a container class for the results of these
+ * calculations and pass this class as the ARGTYPE of both the PerUnitMetricCollector and MultiLevelCollectorReducible.
+ * You can then do these calculations in the makeArg method and they will only be done once per record.
+ *
+ * @param <METRIC_TYPE> The metrics being collected
+ * @param <HISTOGRAM_KEY> If there is are Histograms related to metrics of type <BEAN> then <HKEY> is the key value
+ *                       to these Histograms
+ * @param <ARGTYPE> The type argument passed to individual PerUnitMetricCollector (see SAMRecordMultilevelCollector
+ *                 and PerUnitMetricCollector)
+ * @param <UNIT_COLLECTOR> Class to represent metrics collected at a single level
+ */
+public abstract class MultiLevelCollectorReducible<
+        METRIC_TYPE extends MetricBase,
+        HISTOGRAM_KEY extends Comparable<HISTOGRAM_KEY>,
+        ARGTYPE,
+        UNIT_COLLECTOR extends PerUnitMetricCollector<METRIC_TYPE, HISTOGRAM_KEY, ARGTYPE>
+        >
+    implements Serializable
+{
+    private static final long serialVersionUID = 1L;
+
+    public static final String UNKNOWN = "unknown";
+
+    // A singleton UNIT_COLLECTOR that will accept all records
```

Comment:
Make this a javadoc - here and elsewhere,  all such comments should be in javadoc style